### PR TITLE
Bound MoE routing to kernel capacity

### PR DESCRIPTION
**Pull request by xhjkl**
MoE routing uses fixed threadgroup storage for expert logits, TopK selection, and model input caching. This change centralizes those limits across Rust and Metal and rejects configurations that exceed them before parameter loading or dispatch.

It also validates nonzero expert dimensions, checked hidden-width expansion, and route-count arithmetic without changing the existing MoE execution pipeline.

### CODE REVIEW

File: crates/backend-uzu/build/metal/artifact_cache.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+use std::{collections::HashMap, fs};
```
Comment (CC-Yeh) [inline crates/backend-uzu/build/metal/artifact_cache.rs:1]: Can you elaborate why do we need this exactly?

File: crates/backend-uzu/build/metal/cache_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,321 @@
+use std::{
```
Comment (CC-Yeh) [inline crates/backend-uzu/build/metal/cache_protocol.rs:1]: Same, why do we need this? and this change should not be included in this PR.

File: crates/backend-uzu/tests/unit/encodable_block/moe/moe_block_validation_test.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
+use std::io::Write;
```
Comment (CC-Yeh) [inline crates/backend-uzu/tests/unit/encodable_block/moe/moe_block_validation_test.rs:1]: Don't think we need all these new lines to check very simple logic

**Review by CC-Yeh — CHANGES_REQUESTED** [review]
Can you please remove irrelevant changes?

**Review by CC-Yeh — COMMENTED** [review]
The limits aren't centralized for Metal parts:

- `512` / `128` stay as separate literals in `moe/mod.rs:93,97` and
  `router_topk.metal:9,10`; the diff rewrites the line holding 512 and
  leaves it a literal.
- A const-only `gpu_types/moe.rs` auto-generates `generated/moe.h`
  (pattern: `hadamard_order.rs`, no build wiring)
- The third limit is unbounded: `router_topk.metal:35` declares
  `x_cache[1024]`, `:56` writes `x_cache[c]` for `c < d_model/4` with no
  clamp, and Rust only checks `model_dim % 4 == 0`. `model_dim > 4096`
  overruns a 16 KB threadgroup buffer from a config file.

**Comment by xhjkl** [comment]
Superseded with #746.

File: crates/backend-uzu/build/metal/artifact_cache.rs (hunk 1 above)
Comment (xhjkl) [reply crates/backend-uzu/build/metal/artifact_cache.rs:1]: I'm so sorry: it's indeed an irrelevant change to the PR subject. My intention was to stack those PR together so that you could review or merge then in independently. So I tried to structure them as:
* `build cache` merges into `main`,
* then `mxfp4 groundwork` merges into `build cache`,
* then `proper mxfp4 impl` merges into `mxfp4 groundwork`, and so on the they chain one off another.

But Github apparently won't allow that because I'm not a repo owner of Uzu, so my next PR is a single flat submission, not a chain anymore. Hope this works 🙏
